### DISCUSSION TITLE
Error Handling

### DISCUSSION THREAD

**Discussion by adrianmichalke** [nebulastream/nebulastream#32] [Ideas]
In this discussion, we want to talk about all things error handling.

Just a quick FYI: We are also discussion error handling in the query manager in https://github.com/nebulastream/nebulastream-public/discussions/4#discussioncomment-9882449

**Comment by adrianmichalke** [comment]
### Exception Types


Currently, we use runtime errors a lot (`NES_THROW_RUNTIME_ERROR` ~400 times) to throw exceptions and they are thrown for quite different things. They only differentiate by the attached string. This is a problem because this makes it difficult to handle different errors differently. Most importantly, it makes it more difficult for us (and the user) to reason about errors.

What I would like to have is a set of semantically different exception types on which we agree on. Example list are [here](https://github.com/ClickHouse/ClickHouse/blob/31eeeae4fd8bf416715f3eb9e6a70bbeec4a5deb/src/Common/ErrorCodes.cpp#L16), [here](https://github.com/apple/foundationdb/blob/93b62f6299e841fe4c6bfe9647c01b3d6d18465c/flow/include/flow/error_definitions.h#L34), and [here](https://github.com/duckdb/duckdb/blob/7b276746f772e036474e80f19195a0f5af8891a0/src/include/duckdb/common/exception.hpp#L47).

One way to get to this set would be having an exception type for each way we have to handle an exception (but please think about other ideas):
- `CoordinatorConnectionError` --> catch: log and  retry establish connection
- `InvalidQuerySyntax` --> catch: log and notify coordinator/user about invalid query
- `UnexpectedTuple` --> catch: ignore/log the tuple

For end users we could have these types listed in our doc: [error codes page](https://apple.github.io/foundationdb/api-error-codes.html).

#### Implementation
Other systems implement this is usually as "error codes" in an exception class: An exception class that takes the error code as [enum](https://github.com/duckdb/duckdb/blob/7b276746f772e036474e80f19195a0f5af8891a0/src/include/duckdb/common/exception.hpp) argument. Others embed the error code with more  [pre-compilation](https://github.com/apple/foundationdb/blob/93b62f6299e841fe4c6bfe9647c01b3d6d18465c/flow/include/flow/error_definitions.h) [magic](https://github.com/ClickHouse/ClickHouse/blob/31eeeae4fd8bf416715f3eb9e6a70bbeec4a5deb/src/Common/ErrorCodes.cpp).

**Reply by keyseven123** [reply]
Now that I think about it, we should get rid of the `NES_THROW_RUNTIME_ERROR` and replace it with a similar structure as our logging macros. So that it is quite easy to create different exceptions for the programmers that will then be parsed to the end-user as an error code.

**Reply by adrianmichalke** [reply]
Cool, I think that's how other systems do this. Maybe have a look at the links above. 

What will be the interesting part is to come up with a exception type set that fit to our components.

**Reply by keyseven123** [reply]
Yes, but we could simply start with your three exceptions that you mentioned in this reply and also add one other `QueryExecution` that gets thrown if anything happens during the query execution.  Or is this what you meant by `UnexpectedTuple`?

**Comment by adrianmichalke** [comment]
### Asserts vs. Exceptions

Currently, we are using `NES_ASSERT` in our code base to assert invariants in our code (outside of tests). How should we use and handle them compared to exceptions?

Currently, we just transform a failed assert into a runtime exception.

First idea:
- Asserts should be used to check for conditions that indicate an internal bug or programming error. Use NES_ASSERT to enforce invariants, preconditions, and postconditions that always hold true. Asserts should not be used for handling user input errors, unexpected data, ..., or other recoverable conditions.
-  When an assert fails, NES should shutdown gracefully.

**Reply by keyseven123** [reply]
TBH, I am not sure why we ever would want to use `NES_ASSERT` and not an exception in our code. Except for test code where we might not want to use/create a test exception.

**Reply by adrianmichalke** [reply]
So one advantage that I see is that we can differentiate more between which error types we have. Currently we cannot really 'think' about errors which makes it more difficult to get to a stable system.

Asserts would be a special error type for me. This [article](https://www.methodpark.de/blog/assert-in-production-code-yes-please/) argues that assertions are part of the documentation and a proactive debugging tool, they should only fail if we have messed up something in our code. Exceptions in contrast are used to recover from user/data inputs.
 
 Another option if we want to stick to exceptions would be an `InternalErrorException` that we throw in these cases.

**Reply by zeuchste** [reply]
one point to consider is that as far as I know asserts are not used in the release build only in the debugging

**Reply by keyseven123** [reply]
Just so that I am certain that I understand it by having two examples here:
1. If something happens during the compilation, e.g., a source is not found --> exception
2. If a worker can not allocate a buffer with a given timeout --> assert? 
Or what would be a good example of an assert in our code?

**Reply by ls-1801** [reply]
From my point of view, asserts are invariants that are critical for the function of the code, it should **not** be used to check for runtime errors, unless these errors are explicitly not handled (e.g. out of memory).

If a function is designed to work on certain preconditions (e.g. argument should not be a nullptr) the function should use asserts to verify these conditions. Calling a function with validated preconditions is a critical bug and should immediately terminate the system. In an ideal world, these preconditions are documented and tested.

Runtime errors like timeouts when writing a grpc message or reading a non-existing file should always throw and allow higher-up logic to deal with the problem. This enables modularization, where we can reuse the query compiler in different systems that may want to deal with errors differently. (e.g., the system may have a fallback.

Regarding Asserts in Release builds:
Yes asserts should be disabled in release builds, but this requires that no code with side effects is ever executed within an assert.

**Reply by adrianmichalke** [reply]
> From my point of view, asserts are invariants that are critical for the function of the code, it should **not** be used to check for runtime errors, unless these errors are explicitly not handled (e.g. out of memory).

I would agree on that, eg. timeout should not be handled using asserts but only pre-, postconditions, and invariants of functions.

> one point to consider is that as far as I know asserts are not used in the release build only in the debugging

Im really torn about that one. Maybe the question here is more what we mean when we talk about release builds. If its only about performance yes we should remove them. If it is for some end-user/production environment I would include it.  As a failed assert hints a critical bug to the user it gives more context to the error.

**Reply by zeuchste** [reply]
I think the common thinking here is that the asserts would have been violated during proper testing and that during release execution you do not want to let the system crash rather always have a ordered shutdown with proper error messages

**Comment by kisa7** [comment]
I like the idea of having a list with all exceptions. I also like the idea of every exception to return a code. I feel like if there is a return code it's harder for us to ignore it. My question is, do we want to try to use when possible standart exceptions provided by C++ https://en.cppreference.com/w/cpp/error/exception?

I think another point that we only partially do is to mark all methods that do not throw an exception "noexcept" and make sure we use try and catch if a throw is expected.

**Reply by adrianmichalke** [reply]
Thank you. std::system_error would be the most similar to our central exception as it contains error codes as std::error_codes. But it would be missing the source_location and the stacktrace.

The use of noexcept sounds good. However, I'm not sure if we should mark all functions that don't throw. If I'm not mistaken, any function that triggers an exception even though it is marked with noexcept will terminate the program. Maybe we can follow the C++ core's guidelines that say we should use it when terminating a function due to a throw is impossible or unacceptable. https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#e12-use-noexcept-when-exiting-a-function-because-of-a-throw-is-impossible-or-unacceptable
We could add this to our guidelines?

**Reply by kisa7** [reply]
That make sense!
